x-pack/metricbeat/module/sql/query/query.go

```diff
@@ -0,0 +1,160 @@
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("sql", "query", New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	Driver     string
+	Datasource string
+	QuerySQL   string
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The sql query metricset is beta.")
+
+	config := struct {
+		Driver     string `config:"driver"`
+		Datasource string `config:"datasource"`
+		QuerySQL   string `config:"querySQL"`
+	}{}
+
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		Driver:        config.Driver,
+		Datasource:    config.Datasource,
+		QuerySQL:      config.QuerySQL,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	db, err := sqlx.Open(m.Driver, m.Datasource)
+	if err != nil {
+		return errors.Wrap(err, "error opening connection")
+	}
+
+	err = db.Ping()
+	if err != nil {
+		return errors.Wrap(err, "error testing connection")
+	}
+
+	rows, err := db.Queryx(m.QuerySQL)
+	if err != nil {
+		return errors.Wrap(err, "error executing query")
+	}
+	defer rows.Close()
+
+	// Extracted from
+	// https://stackoverflow.com/questions/23507531/is-golangs-sql-package-incapable-of-ad-hoc-exploratory-queries/23507765#23507765
+	cols, err := rows.Columns()
+	if err != nil {
+		return errors.Wrap(err, "error getting columns")
+	}
+
+	for k, v := range cols {
+		cols[k] = strings.ToLower(v)
+	}
+
+	vals := make([]interface{}, len(cols))
+	for i := 0; i < len(cols); i++ {
+		vals[i] = new(interface{})
+	}
+
+	numericValue := common.MapStr{}
+
+	stringValue := common.MapStr{}
+
+	for rows.Next() {
+		err = rows.Scan(vals...)
+		if err != nil {
+			m.Logger().Debug(errors.Wrap(err, "error trying to scan rows"))
+			continue
+		}
+
+		for i := 0; i < len(vals); i++ {
+			value := getValue(vals[i].(*interface{}))
+
+			num, err := strconv.ParseFloat(value, 64)
+			if err == nil {
+				numericValue.Put(cols[i], num)
+				continue
+			}
+
+			stringValue.Put(cols[i], value)
+		}
```

Comment:
Should we report an event here for each row? If not we will be overwriting the values on each iteration.

Reply:
Changed to report an event for each row, I believe it makes more sense.